### PR TITLE
[release-4.13] WINC-1098: Block upgrades when migrating from in-tree to CSI

### PR DESCRIPTION
**Pull request by sebsoto**
Manual backport of https://github.com/openshift/windows-machine-config-operator/pull/1794

### CODE REVIEW

**Comment by mansikulkarni96** [comment]
@sebsoto does this need to be backported to 4.12?

**Comment by sebsoto** [comment]
@mansikulkarni96 No it does not

File: controllers/controllers.go [hunk 1]
```diff
@@ -110,8 +118,23 @@ func (r *instanceReconciler) ensureInstanceIsUpToDate(instanceInfo *instance.Inf
 	return nc.Configure()
 }
 
+// getNodeSummary queries the cAdvisor metrics endpoint for the given Node
+func (r *instanceReconciler) getNodeSummary(nodeName string) (stats.Summary, error) {
+	var nodeStats stats.Summary
```
Comment (aravindhp) [inline controllers/controllers.go:123]: Please move this closer to the scope where it is used

File: controllers/controllers.go [hunk 2]
```diff
@@ -126,10 +149,15 @@ func (r *instanceReconciler) upgradeBlocked(node *core.Node, upgradingToCSI bool
 	if value := node.GetLabels()[metadata.CSIConfiguredLabel]; value == "true" {
 		return false
 	}
-	if len(node.Status.VolumesAttached) == 0 {
-		return false
+	// For each pod running on this node, check if there is a PVC backed volume, if one is found, block the upgrade
+	for _, pod := range nodeSummary.Pods {
```
Comment (aravindhp) [inline controllers/controllers.go:153]: If you care only about `nodeSummary.Pods` then there is no need to pass the whole object.

File: controllers/controllers.go [hunk 3]
```diff
@@ -110,8 +118,23 @@ func (r *instanceReconciler) ensureInstanceIsUpToDate(instanceInfo *instance.Inf
 	return nc.Configure()
 }
 
+// getNodeSummary queries the cAdvisor metrics endpoint for the given Node
+func (r *instanceReconciler) getNodeSummary(nodeName string) (stats.Summary, error) {
```
Comment (aravindhp) [inline controllers/controllers.go:122]: I suggest just returning `stats.Summary.Pods` given that is all we need and rename the function appropriately. Example: `getPodStatsForNode()`.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @sebsoto

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

Can re-apply LGTM when other comments are addressed

**Review by saifshaikh48 — COMMENTED** [review]
/lgtm

**Review by aravindhp — APPROVED** [review]

File: controllers/controllers.go [hunk 4]
```diff
@@ -110,8 +118,23 @@ func (r *instanceReconciler) ensureInstanceIsUpToDate(instanceInfo *instance.Inf
 	return nc.Configure()
 }
 
+// getPodStatsForNode queries the cAdvisor metrics endpoint for pods on the given Node
+func (r *instanceReconciler) getPodStatsForNode(nodeName string) ([]stats.PodStats, error) {
+	data, err := r.k8sclientset.DiscoveryV1().RESTClient().Get().
+		AbsPath("api/v1/nodes/" + nodeName + "/proxy/stats/summary").DoRaw(context.TODO())
+	if err != nil {
+		return nil, err
```
Comment (alinaryan) [inline controllers/controllers.go:126]: can you add an err msg

**Review by alinaryan — COMMENTED** [review]
just a nit: otherwise lgtm

**Pull request merged**